bsd-user/elfload.c

```diff
@@ -826,6 +830,63 @@ int load_elf_binary(struct bsd_binprm *bprm, struct target_pt_regs *regs,
     return 0;
 }
 
+static void probe_guest_base(const char *filename, target_ulong start,
+                             target_ulong length)
+{
+    struct mapped_range *ranges;
+    unsigned int count;
+    bool overlap;
+    uintptr_t prev, found, host_start, host_end, clamp_start;
+
+    ranges = bsd_get_mapped_ranges(&count);
+    if (ranges == NULL) {
+        fprintf(stderr, "failed to find current mapped ranges\n");
+        exit(-1);
+    }
+
+    prev = qemu_real_host_page_size();
+    overlap = false;
+    found = 0;
+    host_start = guest_base + start;
```

Comment:
Not with -B or reserved_va != 0; this both handles the !have_guest_base case, where it will find free space, and have_guest_base case, where it will give an error if the already-chosen guest_base conflicts with the host's mappings.

Reply:
> Not with -B or reserved_va != 0; this both handles the !have_guest_base case, where it will find free space, and have_guest_base case, where it will give an error if the already-chosen guest_base conflicts with the host's mappings.

Ok. Reverved_va is non zero only on 32 on 64 iirc. This is armv7 and i386 right now (i don't think the powerpc 32bit is working at all right now). I have an armv7 jail already i can test this with when I'm back home later today.... i wonder if this fixes the aarch64 core dump from pkg-static...

Reply:
pkg-static is the case where I first hit things blowing up for riscv64-on-arm64 and motivated my hard-coded guest_base; it's about the only static guest binary you normally end up running during a port build and its base address tends to clash with the host static qemu binary's base address. I should probably just create an armv7 jail locally...